src/wp-admin/includes/plugin.php

```diff
@@ -2596,3 +2596,81 @@ function deactivated_plugins_notice() {
 		update_site_option( 'wp_force_deactivated_plugins', array() );
 	}
 }
+
+/**
+ * Renders an admin notice for each incompatible plugin.
+ *
+ * Displays an admin notice in case a plugin was deactivated during the
+ * loading of WordPress due to incompatibility with the current version
+ * of WordPress.
+ *
+ * @since 6.5.0
+ * @access private
+ *
+ * @global string $wp_version The WordPress version string.
+ */
+function _render_notice_for_deactivated_incompatible_plugins() {
+
+	if ( ! current_user_can( 'activate_plugins' ) ) {
+		return;
+	}
+
+	$blog_deactivated_plugins = get_option( 'wp_force_deactivation_incompatible_plugins' );
+	$site_deactivated_plugins = array();
+
+	if ( false === $blog_deactivated_plugins ) {
+		// Option not in database, add an empty array to avoid extra DB queries on subsequent loads.
+		update_option( 'wp_force_deactivation_incompatible_plugins', array() );
+	}
+
+	if ( is_multisite() ) {
+		$site_deactivated_plugins = get_site_option( 'wp_force_deactivation_incompatible_plugins' );
+		if ( false === $site_deactivated_plugins ) {
+			// Option not in database, add an empty array to avoid extra DB queries on subsequent loads.
+			update_site_option( 'wp_force_deactivation_incompatible_plugins', array() );
+		}
+	}
+
+	if ( empty( $blog_deactivated_plugins ) && empty( $site_deactivated_plugins ) ) {
+		// No deactivated plugins.
+		return;
+	}
+
+	$deactivated_plugins = array_merge( $blog_deactivated_plugins, $site_deactivated_plugins );
+
+	foreach ( $deactivated_plugins as $plugin ) {
+		if ( ! empty( $plugin['version_compatible'] ) && ! empty( $plugin['version_deactivated'] ) ) {
+			$explanation = sprintf(
+			/* translators: 1: Name of deactivated plugin, 2: Plugin version deactivated, 3: Current WP version, 4: Compatible plugin version. */
+				__( '%1$s %2$s was deactivated due to incompatibility with WordPress %3$s, please upgrade to %1$s %4$s or later.' ),
+				$plugin['plugin_name'],
+				$plugin['version_deactivated'],
+				$GLOBALS['wp_version'],
+				$plugin['version_compatible']
+			);
+		} else {
+			$explanation = sprintf(
+			/* translators: 1: Name of deactivated plugin, 2: Plugin version deactivated, 3: Current WP version. */
+				__( '%1$s %2$s was deactivated due to incompatibility with WordPress %3$s.' ),
+				$plugin['plugin_name'],
+				! empty( $plugin['version_deactivated'] ) ? $plugin['version_deactivated'] : '',
+				$GLOBALS['wp_version'],
+				$plugin['version_compatible']
+			);
+		}
+
+		$message = sprintf(
+			'%s</p><p><a href="%s">%s</a>',
+			$explanation,
+			esc_url( admin_url( 'plugins.php?plugin_status=inactive' ) ),
+			__( 'Go to the Plugins screen' )
+		);
+		wp_admin_notice( $message, array( 'type' => 'warning' ) );
+	}
+
+	// Empty the options.
+	update_option( 'wp_force_deactivation_incompatible_plugins', array() );
+	if ( is_multisite() ) {
+		update_site_option( 'wp_force_deactivation_incompatible_plugins', array() );
+	}
```

Comment:
I've been concerned about the impacts to ALL sites of having this code run in these mission critical loader loops. Part of those concerns include:

* All sites will see a performance hit, even if they don't have the Gutenberg plugin activated.
* Writing to the database will also cause a performance hit, though only for those sites with an incompatible Gutenberg plugin version.
* How will it affect caching?

Today, Dion raised the following concerns https://core.trac.wordpress.org/ticket/60457 after the [Plugins Dependency was committed](https://core.trac.wordpress.org/changeset/57545): 

>Running DB Writes, particularly `update_option()` on the front-end, especially during the bootstrap of WordPress can have catastrophic consequences when thousands of requests all try to update the database and cache at the same time.
>
>This leads to a much higher chance of the Database and Cache becoming out-of-sync, which can cause snowballing issues. This happened on api.WordPress.org/plugins today, with WordPress getting stuck in a situation where it constantly tried to update the DB and failed (because the data had already been written there, but the cache said otherwise) and as a result never properly updated the cache.